pkg/guid/guid_pool.go

```diff
@@ -169,55 +176,65 @@ func (p *guidPool) ReleaseGUID(guid string) error {
 func (p *guidPool) AllocateGUID(podUid types.UID, networkId, guid string) error {
 	glog.Infof("AllocateGUID(): podUid %v, networkId %s, guid %s", podUid, networkId, guid)
 
-	if _, err := net.ParseMAC(guid); err != nil {
+	guidAddr, err := net.ParseMAC(guid)
+	if err != nil {
 		return err
 	}
 
 	if !isValidGUID(guid) {
 		return fmt.Errorf("AllocateGUID(): invalid guid %s", guid)
 	}
 
+	guidIndex := guidToUint64(guidAddr)
 	podNetworkId := string(podUid) + networkId
-	if _, exist := p.guidPoolMap[guid]; exist {
+	if _, exist := p.guidPoolMap[guidIndex]; exist {
 		if podNetworkId != p.guidPodNetworkMap[guid] {
 			return fmt.Errorf("failed to allocate requested guid %s, already allocated for %s",
 				guid, p.guidPodNetworkMap[guid])
 		}
 		return nil
 	}
 
-	p.guidPoolMap[guid] = true
+	p.guidPoolMap[guidIndex] = true
 	p.guidPodNetworkMap[guid] = podNetworkId
 	return nil
 }
 
-func (p *guidPool) getNextGUID(currentGUID net.HardwareAddr) net.HardwareAddr {
-	for idx := 7; idx >= 0; idx-- {
-		currentGUID[idx]++
-		if currentGUID[idx] != 0 {
-			break
-		}
+func (p *guidPool) getNextGuidIndex(guidIndex uint64) uint64 {
+	if guidIndex == p.rangeEndIndex {
+		return p.rangeStartIndex
 	}
 
-	return currentGUID
+	return guidIndex + 1
 }
 
-func checkGuidRange(startGUID, endGUID net.HardwareAddr) error {
-	glog.V(3).Infof("checkGuidRange(): startGUID %v, endGUID %s", startGUID.String(), endGUID.String())
+// isValidGUID check if the guild is valid
+func isValidGUID(guid string) bool {
+	match, _ := regexp.MatchString("^([0-9a-fA-F]{2}:){7}[0-9a-fA-F]{2}$", guid)
+	return match && guid != "00:00:00:00:00:00:00:00" && !strings.EqualFold(guid, "ff:ff:ff:ff:ff:ff:ff:ff")
+}
+
+// guidToUint64 convert guid address to uint64 number
+func guidToUint64(guid net.HardwareAddr) uint64 {
+	var result uint64 = 0
+	currentFactor := uint64(math.Pow(0x100, 7))
 	for idx := 0; idx <= 7; idx++ {
-		if startGUID[idx] < endGUID[idx] {
-			return nil
-		}
-		if startGUID[idx] > endGUID[idx] {
-			return fmt.Errorf("invalid guid range. rangeStart: %s rangeEnd: %s", startGUID.String(), endGUID.String())
-		}
+		result += currentFactor * uint64(guid[idx])
```

Comment:
why not do it like this https://github.com/cilium/cilium/blob/master/pkg/mac/mac.go#L52 but in a for loop.
I mean use shift operator<<